### PR TITLE
Ec 755-1

### PR DESCRIPTION
**Pull request by Bromel777**

### CODE REVIEW

**Comment by GusevTimofey** [comment]
change Try(serializer) to Either

File: src/main/scala/encry/consensus/CandidateBlock.scala [hunk 1]
```diff
@@ -11,7 +11,8 @@ case class CandidateBlock(parentOpt: Option[Header],
                           version: Byte,
                           transactions: Seq[Transaction],
                           timestamp: Long,
-                          difficulty: Difficulty) {
+                          difficulty: Difficulty,
+                          stateRoot: Array[Byte]) {
```
Comment (GRYE) [inline src/main/scala/encry/consensus/CandidateBlock.scala]: Возможно нам нужен super tagged тут?😏

File: src/main/scala/encry/local/miner/Miner.scala [hunk 2]
```diff
@@ -88,6 +95,8 @@ class Miner(dataHolder: ActorRef, influx: Option[ActorRef], settings: EncryAppSe
           context.children.foreach(_ ! NextChallenge(candidateBlock))
         case None =>
           logger.info("Candidate is empty! Producing new candidate!")
+          logger.info("send")
+          Thread.sleep(1000)
```
Comment (GRYE) [inline src/main/scala/encry/local/miner/Miner.scala]: 😱

File: src/main/scala/encry/view/state/avlTree/AvlTree.scala [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+      else if (internalNode.key > key) internalNode.leftChild.flatMap(child => getK(key, child))
+      else internalNode.rightChild.flatMap(child => getK(key, child))
+    case leafNode: LeafNode[K, V] => if (leafNode.key === key) Some(leafNode.value) else None
+  }
+
+  def delete(key: K, node: Node[K, V])(implicit m: Monoid[K], v: Monoid[V]): Node[K, V] = balance(delete(node, key).get)
+
+  private def delete(node: Node[K, V], key: K)(implicit m: Monoid[K], v: Monoid[V]): Option[Node[K, V]] = node match {
+    case shadowNode: ShadowNode[K, V] =>
+      val restoredNode = shadowNode.restoreFullNode(storage).get
```
Comment (GusevTimofey) [inline src/main/scala/encry/view/state/avlTree/AvlTree.scala]: You can put it directly into delete function

File: src/main/scala/encry/view/state/avlTree/AvlTree.scala [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+      delete(restoredNode, key)
+    case leafNode: LeafNode[K, V] =>
+      if (leafNode.key === key) None else {
+        deletedNodes = Algos.encode(leafNode.hash) +: deletedNodes
+        Some(leafNode)
+      }
+    case internalNode: InternalNode[K, V] =>
```
Comment (GusevTimofey) [inline src/main/scala/encry/view/state/avlTree/AvlTree.scala]: combine case match pattern with if conditions

File: src/main/scala/encry/view/state/avlTree/AvlTree.scala [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+        val newNode = internalNode.updateChilds(newLeftChild = newLeftChild).selfInspection
+        Some(balance(newNode))
+      } else if (internalNode.key < key){
+        val newRightChild = internalNode.rightChild.flatMap(node => delete(node, key))
+        val newNode = internalNode.updateChilds(newRightChild = newRightChild).selfInspection
+        Some(balance(newNode))
+      } else {
+        val theClosestValue = findTheClosestValue(internalNode, internalNode.key)
+        val hash = implicitly[Hashable[K]]
+        val newNode = theClosestValue match {
```
Comment (GusevTimofey) [inline src/main/scala/encry/view/state/avlTree/AvlTree.scala]: combine this 2 cases

File: src/main/scala/encry/view/state/avlTree/AvlTree.scala [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+
+  private def addToActionInfo(toInsert: HashMap[ByteArrayWrapper, Node[K, V]], toDelete: List[ByteArrayWrapper]): Unit = {
+    //val newDeleted = (deletedNodes ::: toDelete).diff(toInsert.keySet.toList)
+    insertedNodes ++= toInsert
+    deletedNodes :::= toDelete
+  }
+
+  private def balance(node: Node[K, V]): Node[K, V] = node match {
+    case shadowNode: ShadowNode[K, V] =>
+      val restoredNode = shadowNode.restoreFullNode(storage).get
```
Comment (GusevTimofey) [inline src/main/scala/encry/view/state/avlTree/AvlTree.scala]: write it in 1line

File: src/main/scala/encry/view/state/avlTree/AvlTree.scala [hunk 7]
```diff
@@ -0,0 +1,382 @@
...
+      balance(restoredNode)
+    case internalNode: InternalNode[K, V] =>
+      val newAdditionalInfo = (
```
Comment (GusevTimofey) [inline src/main/scala/encry/view/state/avlTree/AvlTree.scala]: you can match it directly w\0 additional val

File: src/main/scala/encry/view/state/avlTree/AvlTree.scala [hunk 8]
```diff
@@ -0,0 +1,381 @@
...
+
+  //return newAvl, allUpdatedNodes and hashes of deleted.
+  def insertAndDeleteMany(version: StorageVersion,
+                          toInsert: List[(K, V)],
+                          toDelete: List[K])
+                         (implicit kSer: Serializer[K],
+                         vSer: Serializer[V],
+                         kM: Monoid[K],
+                         vM: Monoid[V]): (AvlTree[K, V], Map[String, Node[K, V]], List[String]) = {
+    val startTime = System.currentTimeMillis()
```
Comment (GRYE) [inline src/main/scala/encry/view/state/avlTree/AvlTree.scala]: unused

File: src/main/scala/encry/view/state/avlTree/ShadowNode.scala [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+//represent node, that stored in db, without all fields, except hash, height, balance
+case class ShadowNode[K: Serializer: Hashable, V: Serializer](hash: Array[Byte], height: Int, balance: Int)
+                                                             (implicit kM: Monoid[K], vM: Monoid[V]) extends Node[K, V]{
+
+  override val key: K = kM.empty
+  override val value: V = vM.empty
+
+  def restoreFullNode(storage: VersionalStorage): Try[Node[K, V]] = {
+    NodeSerilalizer.fromBytes[K, V](storage.get(StorageKey @@ hash).get)
```
Comment (GRYE) [inline src/main/scala/encry/view/state/avlTree/ShadowNode.scala]: Тут не кинется эксепшен если storage.get вернет None?

File: src/main/scala/encry/view/actors/HistoryApplicator.scala [hunk 10]
```diff
@@ -21,135 +22,161 @@ import encry.view.history.History
...
 
-  def getModifierForApplying(): Unit = if (currentNumberOfAppliedModifiers < setting.levelDB.maxVersions) {
-    logger.debug(s"It's possible to append new modifier to history. Trying to get new one from the cache.")
-    if (locallyGeneratedModifiers.nonEmpty) locallyGeneratedModifiers.dequeueOption.foreach {
-      case (modifier, newQueue) =>
-        locallyGeneratedModifiers = newQueue
+  def getModifierForApplying(history: History): Unit =
+    if (currentNumberOfAppliedModifiers < settings.levelDB.maxVersions) {
+      logger.debug(s"Now it's possible to append new modifier to history. Trying to get new one from the cache.")
+      if (locallyGeneratedModifiers.nonEmpty) locallyGeneratedModifiers.dequeueOption.foreach {
```
Comment (GRYE) [inline src/main/scala/encry/view/actors/HistoryApplicator.scala]: Why would you first check that `locallyGeneratedModifiers` is non empty, and then dequeue option, if you can omit it?

File: src/main/scala/encry/view/actors/NodeViewHolder.scala [hunk 11]
```diff
@@ -34,58 +34,69 @@ import scala.concurrent.{ExecutionContextExecutor, Future}
...
-      }
-    case GetNodeViewChanges(history, state, _) =>
-      if (history) sender() ! ChangedHistory(nodeView.history)
-      if (state) sender() ! ChangedState(nodeView.state)
+      val senderRef = sender()
+      implicit val timeout: Timeout = 30 seconds
+      val newHistory = historyApplicator ? GetHistory
+      val newState = stateApplicator ? GetState
+      val newWallet = walletApplicator ? GetWallet
+      newHistory.foreach(
```
Comment (GRYE) [inline src/main/scala/encry/view/actors/NodeViewHolder.scala]: You can replace it with `for ... yield`

File: src/main/scala/encry/view/state/avlTree/AvlTree.scala [hunk 12]
```diff
@@ -0,0 +1,411 @@
...
+          val newLeaf = LeafNode(newKey, newValue)
+          NodeWithOpInfo(newLeaf, nodeWithOpInfo.opInfo.updateInserted(ByteArrayWrapper(newLeaf.hash) -> newLeaf))
+        }
+        val newNode =
+          internalNode.updateChilds(newRightChild = Some(newRightChild.node), prevOpsInfo = newRightChild.opInfo)
+        balance(newNode.copy(opInfo = newNode.opInfo.update(ByteArrayWrapper(newNode.node.hash) -> newNode.node, ByteArrayWrapper(internalNode.hash))))
+      }
+  }
+
+  private def balance(nodeWithOpsInfo: NodeWithOpInfo[K, V]): NodeWithOpInfo[K, V] = {
```
Comment (GusevTimofey) [inline src/main/scala/encry/view/state/avlTree/AvlTree.scala]: can add @tailrec

File: src/main/scala/encry/view/state/avlTree/AvlTree.scala [hunk 13]
```diff
@@ -0,0 +1,411 @@
...
+      } else {
+        val newRightChild = internalNode.rightChild.map(previousRightChild =>
+          insert(newKey, newValue, nodeWithOpInfo.copy(node = previousRightChild))
+        ).getOrElse{
+          val newLeaf = LeafNode(newKey, newValue)
+          NodeWithOpInfo(newLeaf, nodeWithOpInfo.opInfo.updateInserted(ByteArrayWrapper(newLeaf.hash) -> newLeaf))
+        }
+        val newNode =
+          internalNode.updateChilds(newRightChild = Some(newRightChild.node), prevOpsInfo = newRightChild.opInfo)
+        balance(newNode.copy(opInfo = newNode.opInfo.update(ByteArrayWrapper(newNode.node.hash) -> newNode.node, ByteArrayWrapper(internalNode.hash))))
```
Comment (GRYE) [inline src/main/scala/encry/view/state/avlTree/AvlTree.scala]: You are calling `balance` here with exact same arguments as on line 244, so, you can move it out of if statement and then call

**Pull request closed without merging**